### PR TITLE
Create author profile + befriending logic

### PR DESCRIPTION
**Pull request by olivier-vadiaval**
Create the author profiles
#20 #28 #63 Add befriend, follow, unbefriend logic
Refactor author card by using Django inclusion tags
Refactor delete modal by using Django inclusion tags

### Quick note:
When author A un-befriends author B, it makes more sense that author A is also no longer following author B. I implemented the befriending logic that way. If author A wants to befriend B again, they'll have to follow B. Let me know if that interpretation makes sense.

### Some tips:
- Log in. Must have at least two author in the DB.
- Click on Authors link in nav bar on home page or head to http://127.0.0.1:8000/author/
- Click on your profile card.
   - Check that you cannot befriend yourself. 
   - Check that you can delete your posts from the profile page.
      - Note that it will redirect you to the home page. (Will fix soon)
 - Head back to authors page and befriend other authors and check through the admin site that the friends and followees lists are getting updated appropriately for both the logged-in user and the other author.

---
image
image
image
image

### CODE REVIEW

File: code/socialDistribution/templates/tagtemplates/author.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+                                <a class="px-1" href="{% url 'socialDistribution:unfriend' author.id %}">Un-Friend</a>
+                            {% elif followed %}
+                                <p class="text-muted px-1">Followed &#8226;</p>
+                                <a class="px-1" href="{% url 'socialDistribution:unfollow' author.id %}">Un-Follow</a>
+                            {% elif not author_is_user %}
+                                <a class="px-1" href="{% url 'socialDistribution:befriend' author.id %}">
+                                    {% if accept_friend %}
+                                        Accept friend request
+                                    {% else %}
+                                        Befriend
```
Comment (tylerelston) [inline code/socialDistribution/templates/tagtemplates/author.html:32]: Would it make sense to replace "Befriend" with "Follow" to better reflect the initial follow action?

**Review by tylerelston — APPROVED** [review]
Looks good, made one comment on naming

File: code/socialDistribution/models.py [hunk 2]
```diff
@@ -18,6 +18,59 @@ class Author(models.Model):
...
+            self.followee.remove(author)
+            return True
+        return False
+        
+    def unfriend(self, author):
+        if author.id != self.id and self.is_friends_with(author):
+            self.friend.remove(author)
+            author.friend.remove(self)
+            return True
+        return False
```
Comment (debecerra) [inline code/socialDistribution/models.py:73]: I'm not sure about the data representation of friends/followers. I believe they should be the same thing, it's just that the user stories are intentionally vague, lol. See https://eclass.srv.ualberta.ca/mod/forum/discuss.php?d=1827396#p4797728 for clarification. 

TLDR: A friend request is equivalent to seeking permission to follow. "Friending" someone is the same as "following" someone. A bidirectional follow is a full friendship. 

For the data representation: I think storing both friends and followers will cause issues with data integrity and data duplication. I think we just need one. Maybe we can have a method on Author that checks for friendship by comparing the follow list of both users?

For the UI: I found the flow of actions a bit confusing with the switching around between follow and befriend. What's the difference between unfriend and unfollow? I think we should pick one terminology (i.e. friending or following) and stick to that consistently.

Sorry for the long comment. Hopefully that all makes sense.

File: code/socialDistribution/static/author/style.css [hunk 3]
```diff
@@ -3,13 +3,38 @@
...
+    color: inherit;
+    outline: none;
+}
+
 .author-tile {
     border-radius: 30px;
 }
 
+.author-tile:hover {
+    border-color: lightblue !important;
```
Comment (debecerra) [inline code/socialDistribution/static/author/style.css:23]: Kinda picky, but I'm pretty sure this is a bad practice lol. It's fine though.

File: code/socialDistribution/templates/tagtemplates/author.html (hunk 1 above)
Comment (olivier-vadiaval) [reply code/socialDistribution/templates/tagtemplates/author.html:32]: I agree, it makes more sense. I'll rename it and change it in the mockups.
Thanks for the review ♥

File: code/socialDistribution/models.py [hunk 4]
```diff
@@ -18,6 +18,59 @@ class Author(models.Model):
...
     githubUrl = models.CharField(max_length=50, null=True)
+    followee = models.ManyToManyField('Author', related_name='followees', blank=True)
+    friend = models.ManyToManyField('Author', related_name='friends', blank=True)
+
+    LOCAL = 'LOCAL'
+    REMOTE = 'REMOTE'
+    HOSTS = [
+        ('LOCAL', 'Local'),
+        ('REMOTE', 'Remote')
+    ]
```
[the remark this answers is not in the document]
Comment (olivier-vadiaval) [reply code/socialDistribution/models.py:29]: That makes sense. I misunderstood the logic for dealing with remote authors when working on the ui. I'll make the changes.
Thanks for noticing that ♥

File: code/socialDistribution/static/author/style.css (hunk 3 above)
Comment (olivier-vadiaval) [reply code/socialDistribution/static/author/style.css:23]: Lol yes, sorry, it was so hard to get the border working because of bootstrap. I'll try to fix it though

Comment (olivier-vadiaval) [reply code/socialDistribution/static/author/style.css]: Was able to fix. There was a lingering border-white after the refactoring.

File: code/socialDistribution/models.py (hunk 2 above)
Comment (debecerra) [reply code/socialDistribution/models.py]: Actually, I have an idea. Instead of folloewee/friends, we can have a **follower** field. 

A sends a follow/friend request to B. This is not stored in the Authors db yet, it just goes to B's inbox. Once they accept, A gets added to their "followers". This way, when A makes a post, we send it to Bs inbox since we know B is a follower. If B has A in its followers and B has A in its followers, they are "full friends"

Let me know what you think.

**Review by JawadHossain — APPROVED** [review]
Nice!

**Comment by debecerra** [comment]
Hey @olivier-vadiaval, how does someone accept a request?

**Comment by debecerra** [comment]
Great, I'll approve for now, but before you merge would you be able to make a new issue or two with the outstanding tasks regarding inbox and following? Just want to make sure nothing slips through the cracks. I'll do the same for any outstanding items I can think of.

**Review by debecerra — APPROVED** [review]

**Comment by olivier-vadiaval** [comment]
> Overall, I think it looks pretty good. It addresses what we talked about.
> 
> We'll have to readdress the inbox again later when we start to consider remote authors. It won't work the way it's set up right now. Also might be worth redesigning the inbox in some way. For example, posts need to stay saved on the server, but friend requests are more like a queue, so once they are accepted the request can be discarded. Either way, this appears to address local follow functionality which is great. Once you tell me how to actually accept a request, I'll approve.

You mean like delegating some of the inbox logic to the model itself instead of the views? In the changes I made, the view functions are handling the follow/un-follow/friend logic. Once a user accepts a request, the view removes the association between that user and the follower for the follow_request field and then adds that person as a follower.
